src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamListener.cs

```diff
@@ -66,22 +65,40 @@ public override async Task PollAsync(CancellationToken cancellationToken)
             IDatabase db = multiplexer.GetDatabase();
             StreamEntry[] entries = await db.StreamReadGroupAsync(key, consumerGroup, consumerName, count: count);
             logger?.LogDebug($"{logPrefix} Received {entries.Length} entries from the stream at key '{key}'.");
-            await Task.WhenAll(entries.Select(entry => ExecuteAsync(entry, cancellationToken)));
+            if (entries.Length == 0)
+            {
+                return;
+            }
+
+            if (batch)
+            {
+                await ExecuteAsync(entries, Array.ConvertAll(entries, entry => entry.Id), cancellationToken);
+            }
+            else
+            {
+                await Task.WhenAll(entries.Select(entry => ExecuteAsync(entry, new RedisValue[] { entry.Id }, cancellationToken)));
+            }
         }
 
-        private async Task ExecuteAsync(StreamEntry entry, CancellationToken cancellationToken)
+        private async Task ExecuteAsync(object value, RedisValue[] entryIds, CancellationToken cancellationToken)
         {
             IDatabase db = multiplexer.GetDatabase();
-            await executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = entry }, cancellationToken);
+            FunctionResult result = await executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = value }, cancellationToken);
 
-            RedisValue[] entryIds = new RedisValue[] { entry.Id };
-            long acknowledged = await db.StreamAcknowledgeAsync(key, consumerGroup, entryIds);
-            logger?.LogDebug($"{logPrefix} Acknowledged {acknowledged} entries from the stream at key '{key}'.");
+            if (result.Succeeded)
+            {
+                long acknowledged = await db.StreamAcknowledgeAsync(key, consumerGroup, entryIds);
```

Comment:
Again, not directly related to batching - I noticed that the acknowledgement only happens on the success path. What will happen if the user has some bad code in their function that always throws an exception and never completes successfully? Does this stream element get retried and if so how many times? In systems such as service bus, there is a setting on the queue that controls how many attempts at delivery will be made before the message is dead lettered - this prevents infinite retry and I'm curious about what is expected in the redis case.

Reply:
Neither Redis Streams nor Redis Lists have settings that control the attempts at delivery. Redis Lists don't have a method for acknowledging to Redis that an element has been successfully processed either, as polling a list requires popping elements from the list, meaning that each element can be retrieved once only. Redis Stream elements can be [acknowledged](https://redis.io/commands/xack/) so that the element is not returned to the caller on subsequent calls. [Although the server-side keeps track of how many times each element has been sent to a client](https://redis.io/docs/data-types/streams/#consumer-groups), there is no automatic disposing of elements after a threshold, and it is fully up to the client to acknowledge the element.

I've simplified the PR to focus specifically on the batching logic, which means reverting to the current single delivery attempt for both the list and stream triggers, but I do believe that we should add a "maxRetries" setting for customers in the future.